### ISSUE TITLE
Offer safe variants for the RIO.List.Partial functions in the RIO.List module

### ISSUE THREAD

**Issue by roman** [commercialhaskell/rio#82]
Iterating over the work from #43, what do you think about adding all the functions from `RIO.List.Partial` with a `safe` prefix in the `RIO.List` module?

There is no clear replacement for the functions in `RIO.List.Partial`, and I would avoid them as much as possible if I can.

**Comment by snoyberg** [comment]
In principle: I'm OK with this approach. I'm not convinced that `safe` will be the right prefix in all cases. For example: `maximum` would probably be better replaced by `maximumMay` or `maximumMaybe` to match existing naming conventions IMO.

**Comment by mitchellwrosen** [comment]
FYI: `scanl1`, `scanr1` aren't partial

**Comment by snoyberg** [comment]
Implemented in #96

**Comment by tejon** [comment]
Late entry, but perhaps relevant... yesterday I had to switch some lists over to vectors so I could use `statistics` operations on them. I had been using some of the (list-specific) `*Def` functions from `safe`, and needed to replicate those; and some of the `statistics` operations I wanted were partial on empty vectors, too, which needed fixing. I wound up writing this:

```haskell
defaulting :: Foldable f => b -> (f a -> b) -> f a -> b
defaulting d f xs
  | null xs   = d
  | otherwise = f xs
```

So now I have e.g. `defaulting 0 Vector.maximum`. The `*Maybe` pattern is a trivial special case:

```haskell
mayhap :: Foldable f => (f a -> b) -> f a -> Maybe b
mayhap f = defaulting Nothing (Just . f)
```

Are there any particular downsides to this technique? It seems far more elegant than a slew of specialized, type-specific functions.
